### DISCUSSION TITLE
Add option for busy wait frame limiting

### DISCUSSION THREAD

**Discussion by theyareonit** [ppy/osu#27547] [Ideas] [answered]
I decided to throw together a busy wait based frame limiter to see how significant the performance difference would be compared to the default WaitableTimer frame limiter.
![image](https://github.com/ppy/osu/assets/82268986/c688a987-8d0e-4212-aa04-89a5832705b3)

Testing was done on single-threaded mode, DX11, using the default 1ms timer resolution.
![image](https://github.com/ppy/osu/assets/82268986/16b0a27c-9f62-4955-9de4-929bfdacaf81)
The % lows aren't hugely different, because they're bottlenecked by unrelated frametime spikes, but notice that the standard deviation is more than halved with the busy wait frame limiter. Using a [more precise timer resolution](https://github.com/amitxv/PC-Tuning/blob/main/docs/research.md#micro-adjusting-timer-resolution-for-higher-precision) improves things, but still performs significantly worse than the busy wait limiter.

Downsides:
- Battery life
- Will starve background tasks of resources on low core count systems
- Helps much more in single-threaded mode than multi-threaded mode, since busy waiting every single thread would be a bit overboard

It's also possible to do a hybrid of busy waiting and sleeping, by sleeping for the first part of a frame and then busy waiting the rest. However, the author of [Special K](https://www.special-k.info/) recommends sleeping for at most `floor(frametime - (2.5 * timer resolution))ms` to prevent harming performance, which would mean pure busy waiting above 286fps anyway on osu!'s default 1ms timer resolution.

**Comment by peppy** [comment]
this is pretty much a "no" from me, already tested and deemed it to be fruitless waste of processing power. but there are other options going forward like better graphics apis for frame sync.

what are the numbers in unit wise in your reported stats?

**Reply by theyareonit** [reply]
The unit is 1/frametime (relative to each previous frame, not an average)

Here's a more visual comparison if it helps, the numbers don't really represent what's happening very well:

WaitableTimer
![image](https://github.com/ppy/osu/assets/82268986/8bfdace7-77ae-4fee-abe6-2e2d516a3a88)

Busy wait
![image](https://github.com/ppy/osu/assets/82268986/dd933a97-b78d-4ba0-a357-9849134ff72d)

I don't really see the harm in just adding it as an extra option, since it definitely does improve performance as long as you have enough cores, and it'll improve performance much more as these larger spikes get ironed out.

**Reply by peppy** [reply, marked as answer]
In the interest of reducing variables, and the environment impact (however small it is), I am against this. Your graphs affirm my decision, as far as I'm concerned.

Sorry if this doesn't align with your visions, but it's a call I've already made.

**Comment by sineplusx** [comment]
@theyareonit Do you have the source code for this uploaded somewhere? I would like to run a few tests with it as well.

**Reply by theyareonit** [reply]
Sorry for taking so long to respond, basically just replace the `throttle()` function in https://github.com/ppy/osu-framework/blob/master/osu.Framework/Timing/ThrottledFrameClock.cs with this:
```
        private void throttle()
        {
            double targetFrameTime = 1000d / MaximumUpdateHz;

            // The modulo term corrects for lag-induced desync
            while (SourceTime - (CurrentTime - (CurrentTime % targetFrameTime)) < targetFrameTime)
            {
            }
            CurrentTime = SourceTime;
        }
```
It might make you unable to hit high FPS values (>700), since this frame limiter is more eager to drop frames than vanilla (though to be honest I think this may result in a smoother experience, since lagspikes will just mean individual dropped frames rather than dozens of frames of "catchup").

Lazer's framepacing is still significantly worse than stable on my machine, so a change in frame limiter is likely to make much more of a difference once overall performance is improved.
